test/test-define-x-expander-use-site-scope.rkt

```diff
@@ -0,0 +1,73 @@
+#lang racket
+
+(require syntax/parse
+         syntax/parse/experimental/eh
+         generic-syntax-expanders
+         syntax/stx
+         (for-syntax syntax/parse
+                     racket/syntax)
+         rackunit)
+
+(define-expander-type eh-mixin)
+
+(begin-for-syntax
+  (define eh-post-accumulate (make-parameter #f)))
+
+(define-for-syntax (strip-use-site stx)
+  (define bd
+    (syntax-local-identifier-as-binding (syntax-local-introduce #'here)))
+  (define delta
+    (make-syntax-delta-introducer (syntax-local-introduce #'here) bd))
+  (delta stx 'remove))
+
+(define-syntax define-eh-alternative-mixin
+  (syntax-parser
+    [(_ name ((~literal pattern) pat) ... (~optional (~seq #:post post)))
+     (let ()
+       #`(define-eh-mixin-expander name
+           (λ (_)
+             #,@(if (attribute post)
+                    #`(((eh-post-accumulate) (quote-syntax post)))
+                    #'())
+             (quote-syntax (~or . #,(strip-use-site #'(pat ...)))))))]))
+
+(define-syntax ~no-order
+  (pattern-expander
+   (λ (stx)
+     (syntax-case stx ()
+       [(self pat ...)
+        (let ()
+          (define acc '())
+          (define (add-to-acc p)
+            (set! acc (cons p acc)))
+          (define alts
+            (parameterize ([eh-post-accumulate add-to-acc])
+              (expand-all-eh-mixin-expanders
+               #'(pat ...))))
+          #`(~and (~seq (~or . #,alts) (... ...))
+                  #,@acc))]))))
+
+;; Test:
+
+(define-eh-alternative-mixin aa
+  (pattern (~optional (~and some-pat #:some))))
+
+(define-eh-alternative-mixin bb
+  (pattern (~optional (~and other-pat #:other)))
+  ;; Without the fix in PR #8, the following line gives the error
+  ;;   attribute: not bound as a pattern variable in: some-pat
+  #:post (~fail #:when (and (attribute some-pat)
```

Comment:
I think that's it. I'll rephrase to double-check we understand the same thing:

* When `define-?-expander` is called, racket automatically adds a "use-site" scope to all of its arguments, as it thinks the macro is likely to produce an expression.
* `define-?-expander` produces a `(define-syntax expander-name body)` definition (not expressions), which, to the best of my knowledge, do not need that extra "use-site" scope. To be 100% percent sure we'd have to ask @mflatt.
* Any `(quote-syntax xyz)` or `#'xyz` supplied by the user to `define-?-expander` will have this scope attached to the whole form. When the `expander-name` (in this case `aa` and `bb`) defined that way executes, the extra scope is therefore present on the `xyz` piece of syntax.
* In the test code, `(~no-order (aa) (bb))` gets expanded to:
```
(~and (~seq (~or (~optional (~and some-pat #:some))
                 (~optional (~and other-pat #:other)))
            ...)
            (~fail #:when (and (attribute some-pat)
                               (attribute other-pat))))
```
* From a "scope" point of view (as in inside or outside of the scope of a `let`), this roughly has the structure:
```
(let [(some-pat blah)
      (other-pat blahblah)]
  (and some-pat other-pat))
```
* However, due to the extra scope, and since the `(and (attribute some-pat) (attribute other-pat))` expression is injected by `bb`, it becomes:
```
(let [(some-pat¹ blah¹)
      (other-pat² blahblah²)]
  (and some-pat² other-pat²))
```
* In other words, all syntax objects produced by `aa` have the scope `¹`, and all syntax objects produced by `bb` have the scope `²`.
* In my test example, the `some-pat` and `other-pat` code are directly injected into the macro's code and ultimately come from the user, so they behave as if the macro created the pieces of syntax and passed them to `syntax-local-introduce`, which should make them unhygienic (i.e. the two macros can communicate). Still, this is not enough, as there is this extra scope.
* Since the extra scope is added by `define-?-expander` on the whole body of `aa` and `bb`, they can't remove it. In other words, it's not a new fresh scope added for each invocation of the expander, which could be removed with `syntax-local-introduce` or `syntax-local-identifier-as-binding`, but one present on their body, across all invocations (if I understood correctly).
* Exactly as you say, it's very hard to remove that scope inside the macro definition, so it would be better to just not add it.

Scopes are a very complex topic, so I hope I analysed the situation well.

While writing this I understood the problem better, and managed to craft a much simpler test case, which I'll push soon.